platform/http_session_manager.mm

```diff
@@ -58,6 +59,8 @@ - (instancetype)init
                                         delegateQueue:nil];
     _taskInfoByTaskID = [NSMutableDictionary dictionary];
     _taskInfoQueue = dispatch_queue_create("http_session_manager.queue", DISPATCH_QUEUE_CONCURRENT);
+    BOOL const isSyncHttpTest = [NSBundle.mainBundle.executableURL.lastPathComponent hasSuffix:@"osm_auth_tests"];
```

Comment:
It looks strange.

Is it possible to detect that we are already in _main queue_ while performing ```dispatch_async(dispatch_get_main_queue()```
I suppose that ObjC core should do it itself, something like _coroutine_ concept.

Reply:
The main thread can be detected by `NSThread.isMainThread`, but it doesn't help here. The delegate is called from another thread. Old code posted a block to the main thread that was blocked in the test by `dispatch_group_wait(group, DISPATCH_TIME_FOREVER);` so the posted block was never executed.

Reply:
Ok then, if we already have taskInfoQueue for processing task infos, why we can't use it for running completionHandler always? Especially, because we use taskInfo.delegate and task itself to run handler.

Otherwise, assume calling dispatch taskInfo.delegate to dispatch_get_main_queue and calling ```removeTaskInfoForTask``` at the same time?

Reply:
Because Apple's dispatch queues are tricky. There is no guarantee that any queue except the main one will run on the same thread. The same queue can reuse different threads depending on their availability. And we have a lot of thread checkers in chunker downloader for example to make sure that its callbacks are always single-threaded.

> Work submitted to dispatch queues executes on a pool of threads managed by the system. Except for the dispatch queue representing your app's main thread, the system makes no guarantees about which thread it uses to execute a task.

[source](https://developer.apple.com/documentation/dispatch/dispatchqueue)

Reply:
This all mess should be rewritten properly using thread pools and lambdas )